Depth supervision for nerfacto

mpmisko opened the pull request:
Implements depth-supervised Nerfacto based on with depth losses from [Depth-supervised NeRF](https://arxiv.org/abs/2107.02791) and [Urban Radiance Fields](https://arxiv.org/abs/2111.14643). I found the [Depth-supervised NeRF](https://arxiv.org/abs/2107.02791) loss to perform better for my dataset, however, [Urban Radiance Fields](https://arxiv.org/abs/2111.14643) reports it is SOTA, so it may work better for other use cases. I set the [Depth-supervised NeRF](https://arxiv.org/abs/2107.02791) loss as default with some reasonable initial hyperparameters. 

I tested the method in an indoor livingroom scene. The additional depth loss helps with depths (see depth_mse), however slightly hurts other eval metrics:

<img width="1090" alt="Screen Shot 2023-01-01 at 10 42 22 PM" src="https://user-images.githubusercontent.com/18229042/210185409-f0cd4be9-17f9-4065-9299-24e2832e5179.png">

Example depth colormaps (left is nerfacto & right is nerfacto with depth supervision, left image is ground truth depth in both cases).

<img width="1302" alt="Screen Shot 2023-01-01 at 10 44 33 PM" src="https://user-images.githubusercontent.com/18229042/210185489-99a218fc-00fa-4e46-bed9-543378f87df1.png">

I generally find that the additional depth supervision removes floaters or makes the floaters more diffuse in the worst case. However, it seems to me that nerfacto is slightly sharper around objects, which could be explained by noise in depth measurements. Depth supervision will be probably be helpful on a case-by-case basis.

tancik commented on nerfstudio/cameras/camera_utils.py:
```diff
@@ -230,6 +230,18 @@ def normalize(x) -> TensorType[...]:
+
+    Args:
+        x: tensor to normalize
+        dim: axis along which to normalize
```
Add a `Returns: ` to docstring

tancik commented on nerfstudio/data/dataparsers/nerfstudio_dataparser.py:
```diff
@@ -298,3 +284,20 @@ def _get_fname(self, filepath: PurePath, data_dir: PurePath, downsample_folder_p
         return data_dir / filepath
+
+
+def get_indices_for_split(split: str, num_images: int, train_split_percentage: float) -> np.ndarray:
```
Add documentation

tancik commented on nerfstudio/data/dataparsers/depth_dataparser.py:
```diff
@@ -0,0 +1,61 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Depth data parser."""
```
Given that this is following the same format at the nerfstudio data, what are your thoughts on adding a `if "depth_file_path" in frame` logic to the nerfstudio-dataparser (similar to what we do with masks)

mpmisko replied:
Great idea! Went went this approach instead.

tancik commented on nerfstudio/data/datasets/depth_dataset.py:
```diff
@@ -0,0 +1,48 @@
+    """Dataset that returns images and depths.
+
+    Args:
+        dataparser_outputs: description of where and how to read input images.
```
Add `Returns:`

mpmisko replied:
Sorry, not totally sure what this refers to as this comment relates to` __init__`.

tancik replied:
Oops

tancik commented on nerfstudio/data/utils/data_utils.py:
```diff
@@ -51,3 +54,17 @@ def get_semantics_and_mask_tensors_from_path(
+    scene_scale_factor: float,
+    interpolation: int = cv2.INTER_NEAREST,
+) -> torch.Tensor:
+    """Loads, rescales and resizes depth images. Assumes Record3D depth format."""
```
Add `Args:` and `Returns:`

tancik commented on nerfstudio/data/utils/data_utils.py:
```diff
@@ -51,3 +54,17 @@ def get_semantics_and_mask_tensors_from_path(
+) -> torch.Tensor:
+    """Loads, rescales and resizes depth images. Assumes Record3D depth format."""
+    image = cv2.imread(str(filepath.absolute()), cv2.IMREAD_ANYDEPTH)
+    image = image.astype(np.float64) * _MILLIMETER_TO_METER_SCALE_FACTOR * scene_scale_factor
```
Should this scale factor be a variable? If not, it should be well documented.

mpmisko replied:
Good point, I made it a variable instead in the `NerfstudioDataParserConfig`.

tancik commented on nerfstudio/model_components/losses.py:
```diff
@@ -194,3 +193,31 @@ def pred_normal_loss(
+    lengths: TensorType[..., "num_samples", 1],
+    sigma: TensorType[0],
+) -> TensorType[..., 1]:
+    """Depth loss from Depth-supervised NeRF (Deng et al., 2022)."""
```
Add full docs

tancik commented on nerfstudio/model_components/losses.py:
```diff
@@ -194,3 +193,31 @@ def pred_normal_loss(
+    sigma: TensorType[0],
+    directions_norm: TensorType[..., 1],
+    is_euclidean: bool,
+) -> TensorType[0]:
```
Add full docs

tancik reviewed (commented):
Thanks for working on this! I added a few comments, mostly on missing documentation. Two additional things, 1. can you provide some example results in this PR, ie a scene trained with and without DS to verify that it is working as expected. 2. Can you add some documentation with respect to the expected format of the depth images. This can be added to some of the docstrings, but also it would make sense to add it to this page - https://github.com/nerfstudio-project/nerfstudio/blob/main/docs/quickstart/data_conventions.md
Specifics such as, num channels, data format (uint8 vs 16 bit), is it "distance" vs "depth", what resolution it should be, how to account for sparsity (from lidar), how to specify path in json.

mpmisko commented:
Thanks for the nice review @tancik!

I believe I addressed all of the comments. Please take a look when you can.

I also added some results to the PR description and updated the docs. Happy to elaborate more if you have some comments.

tancik reviewed (approved):
LGTM! Thanks for making the changes

penguinflys commented:
@mpmisko Thanks for the excellent coding w.r.t depth observation, can you please give the reference where the depth sigma decaying origins from, or maybe you created it?

YueYin27 commented:
@mpmisko Hi, I was trying depth-nerfacto with my blender dataset, but it raises errors. Does depth-nerfacto only support the nerfstudio dataset? Thank you!

pushkalkatara commented:
Hi @mpmisko , Thanks for the amazing contribution. 

Could you please provide the dataset/scene you used for evaluation?

In my case, I'm using a scene from Scannet:

depth loss increases, depth visualizations look okay, but not great. 
holes seem to fill but the depth scale doesnt seem right. [bluish tint]

<img width="389" alt="image" src="https://github.com/nerfstudio-project/nerfstudio/assets/21266230/0c454070-7628-45c6-8b7a-b86b20cdea6b">


Unprojections: left [using nerf depth] , right [sensor depth]
![image](https://github.com/nerfstudio-project/nerfstudio/assets/21266230/8e5e08e5-68d2-437e-a79d-6b228be2246c)
